### ISSUE TITLE
Dynamic AI information about your deck

### ISSUE THREAD

**Issue by ShawnieBoy** [magarena/magarena#518]
Just some musing about the AI and it's known information, and a way of allowing the AI to benefit from 'Reveals' and potential 'Look at opponent's hand' effects.

Currently the AI has full knowledge of the contents of your deck, although not necessarily where any of those cards are. It makes assumptions about the order of your deck and the contents of your hand given your full decklist.

I am wondering if the AI could adjust the information it knows more dynamically as play progresses.

At the start of the game it could assume that your deck contains lands, but no other cards. So your deck would be filled with a pseudo card -
```
name=Land
rarity=B
type=Basic,Land
timing=land
ability={T}: Add {1} to your mana pool.
oracle={T}: Add {1} to your mana pool.
```
As play progresses, and cards are played or revealed, a potential version of the players deck is 'built'.
This would be a decklist of known cards, supplemented with assumed cards.

Continuing from the start of the game, you play a Forest. The AI could then replace all the assumed pseudo-cards with Forests.
You then cast a Wall of Wood. The AI now knows that one, and can make an assumption that there are, at most, another three of them in the deck, so adds three Wall of Wood to the assumed list, and one to the known list.
The assumed list would replace the most populous card first, and when one is added to the known list, the assumed list is reduce by one.

As play progresses the decklist will be revealed.

The reason for all this would be allowing the function of revealing your hand, or looking at your opponents hand (or library etc.). These cards can then be added to the decklist. Also the hand can then also become a known zone, similarly built as the decklist.

Simulations would work with known information of the hand, and as cards are played from that hand, removed and replaced with assumed cards. (Or more known cards if the top of the library is revealed etc.)

This is all 'blue sky' thinking, but is the only way I can see that the AI could benefit from being shown cards.

**Comment by firemind** [comment]
Interesting topic. If you think the AI could benefit from making smart assumptions about the content of the opponents deck, I might have a suggestion here.

I think there are two main scenarios in which a real player would make reasonable assumptions about the opponents deck:

1. There are fixed decks from which players pick. If I see a couple of cards I might figure out which one of the pre-built decks my opponent is playing. This applies to Magarena in that the AI could "narrow it down" to the available .dec files.

2. We're playing a format like Modern and I can figure out what archetype my opponent is playing given a couple of key cards. For example: if my opponent plays turn one Blinkmoth Nexus into Springleaf Drum it is highly likely that they are playing Affinity. Operating under the assumption that my opponent has Cranial Plating and Arcbound Ravager in their deck would thus be reasonable. So if the AI could "figure out" what archetype the opponent is playing it could use a stock list of that archetype to "fill in the blanks" in the opponents deck list. I'm bringing this up because I've been experimenting with classification algorithms (mainly Random Forests) to improve the archetype system for firemind.ch and it shows that you can get a pretty accurate guess at a deck list from just a few cards.

I realize that archetypes only apply to a fraction of games with Magarena but it seems like the only place where "smart assumptions" are even possible. There is also the more volatile topic of draft/sealed archetypes which are harder to classify. Still, a somewhat accurate guess seems better than no guess at all.

**Comment by melvinzhang** [comment]
There is a potential problem with MCTS, at the start (when it has minimal knowledge) it would observe in its simulations that the opponent never wins (it would be very hard to win with a deck full of lands) and almost any moves it makes results in it winning. This generally results in bad play by the MCTS AI.

This is the main reason why the honest version of MCTS AI appeared two years after the cheating version of MCTS AI. The original honest version (never released) assumed that all unknown cards are {15} mana cost 1/1 creatures and it plays very badly, I think using lands would be similar. The solution turns out to be the assuming a random ordering of the contents of the deck and a random selection of the deck as the hand. 

There is a subtlety in constructing the hand for simulations. Cards which are known to be in the hand, such as those reveals when adding to hand, must stay in the hand. This is to allow Miracle trigger to work correctly, without this the Miracle card that was in opponent's hand would be replaced.

**Comment by ShawnieBoy** [comment]
If the opponents hand already takes into account revealed cards, would the whole 'look' and 'reveal' with regards to the hand be able to be implemented using that process? Such as 'look at target players hand' or 'players play with their hands revealed'?

**Comment by melvinzhang** [comment]
Yes, that was always possible mechanically. The blocking issue is that the AIs are not able to utilize these effects properly, that is why they were not implemented, except for cases of reveal which were needed as part of other effects such as search to hand. 

As you mentioned, the problem is that the MCTS AI does not realize the benefit of using these effects.
